galvan/src/main/java/org/terracotta/testing/master/GalvanStateInterlock.java

```diff
@@ -339,50 +358,52 @@ public void forceShutdown() throws GalvanFailureException {
       this.logger.output("> forceShutdown");
       // Set the flag that we are shutting down.  That way, any servers which were concurrently coming online can be stopped when they check in.
       this.isShuttingDown = true;
+    }
 //  trying to debug a Galvan hang where not all the servers are seen
       long timeExpired = System.currentTimeMillis() + TimeUnit.MINUTES.toMillis(1);
       // We wait until there is no active, no passives, no unknown servers, and no running clients.
-      while (timeExpired > System.currentTimeMillis() && (
-          (null != this.activeServer)
-          || !this.passiveServers.isEmpty()
-          || !this.unknownRunningServers.isEmpty()
-          || !this.runningClients.isEmpty()
-        )
-      ) {
-        this.logger.output("* forceShutdown waiting on active: " + (null != this.activeServer)
-            + " passives: " + this.passiveServers.size()
-            + " unknown: " + this.unknownRunningServers.size()
-            + " clients: " + this.runningClients.size()
-            );
+      while (timeExpired > System.currentTimeMillis() && !checkIfEmpty()) {
+        synchronized (this.sharedLockState) {
+          this.logger.output("* forceShutdown waiting on active: " + (null != this.activeServer)
+              + " passives: " + this.passiveServers.size()
+              + " unknown: " + this.unknownRunningServers.size()
+              + " clients: " + this.runningClients.size()
+              );
+        }
         // Force shut-down all clients, all passives, and the active.
         // (note that we won't modify the collections here, just walk them - we are synchronized)
         // this is inside the loop because servers can transition from terminated to active or passive 
         // when the lock is released during the wait, if access is at all fair, should not cause double
         // shutowns
-        for (ClientRunner client : this.runningClients) {
+        for (ClientRunner client : makeCopy(this.runningClients)) {
           client.forceTerminate();
         }
-        for (ServerProcess server : this.unknownRunningServers) {
+        for (ServerProcess server : makeCopy(this.unknownRunningServers)) {
           safeStop(server);
         }
-        for (ServerProcess server : this.passiveServers) {
+        for (ServerProcess server : makeCopy(this.passiveServers)) {
           safeStop(server);
         }
-        if (null != this.activeServer) {
-          safeStop(this.activeServer);
+        ServerProcess active = null;
+        synchronized (this.sharedLockState) {
+          active = this.activeServer;
+        }
+        if (null != active) {
+          safeStop(active);
         }
         safeWaitWithTimeout(10, TimeUnit.SECONDS);
```

Comment:
This call fails now as no lock is held during the invocation
 ```
 private void safeWaitWithTimeout(long time, TimeUnit units) {
    try {
      this.sharedLockState.wait(units.toMillis(time));
    } catch (InterruptedException e) {
      Assert.unexpected(e);
    }
  }
```

Reply:
fixed